### PR TITLE
initial logging details tree

### PR DESCRIPTION
**Pull request by pq**
image

/cc @jacob314

### CODE REVIEW

File: lib/inspector/diagnostics_node.dart [hunk 1]
```diff
@@ -23,23 +26,31 @@ const Map<String, DiagnosticLevel> diagnosticLevelNames = {
...
+  'flat': DiagnosticsTreeStyle.flat,
   'singleLine': DiagnosticsTreeStyle.singleLine,
+  'headerLine': DiagnosticsTreeStyle.headerLine,
   'indentedSingleLine': DiagnosticsTreeStyle.indentedSingleLine,
   'shallow': DiagnosticsTreeStyle.shallow,
   'truncateChildren': DiagnosticsTreeStyle.truncateChildren,
 };
 
+final Map<DiagnosticsTreeStyle, String> treeStyleToName = _invertMap(nameToTreeStyle);
+
```
Comment (devoncarew) [inline lib/inspector/diagnostics_node.dart]: nit: I'd run this through dartfmt

File: lib/inspector/diagnostics_node.dart [hunk 2]
```diff
@@ -345,7 +356,10 @@ class RemoteDiagnosticsNode extends DiagnosticableTree {
     if (value == null) {
       return defaultValue;
     }
-    final level = diagnosticLevelNames[value];
+    final level = nameToDiagnosticLevel[value];
+    if (level == null) {
+      print ('Unabled to find level for $value');
```
Comment (devoncarew) [inline lib/inspector/diagnostics_node.dart]: We should fail in some other way than `print` here. Perhaps just default to some diagnostic level?

File: lib/logging/logging.dart [hunk 3]
```diff
@@ -578,12 +608,57 @@ class LogDetailsUI extends CoreElement {
...
+            tree.maybePopulateChildren(node);
+          }
+          // The ui wasn't updated because this selection was triggered by the logging
+          // view instead of the regular inspector.
+          node.diagnostic.setSelectionInspector(false);
+          // TODO(jacobr): warn if the selection can't be set as the node is
+          // stale which is likely if this is an old log entry.
+        },
+        onExpand: (e) {
+          print('XXX expand $e');
```
Comment (devoncarew) [inline lib/logging/logging.dart]: Perhaps add a todo to implement later (and remove the `print`)?

File: lib/ui/fake_flutter/diagnosticable.dart [hunk 4]
```diff
@@ -7,6 +7,9 @@
 // with a couple of tweaks to support error levels that haven't yet landed in
 // the core Flutter repo but should land soon.
 part of 'fake_flutter.dart';
+// Copyright 2017 The Chromium Authors. All rights reserved.
```
Comment (devoncarew) [inline lib/ui/fake_flutter/diagnosticable.dart]: Is the dupe of the copyright comment intended?

**Comment by devoncarew** [comment]
Is your screenshot up to date? I don't see the code to create the reload and restart button in  logging.dart (and generally, think that we shouldn't put them in every page, but in some future PR create a area in the UI that's global to the app, which can hold those buttons).

File: lib/inspector/diagnostics_node.dart (hunk 2 above)
Comment (jacob314) [inline lib/inspector/diagnostics_node.dart]: remove this debugging print and instead add a string message to the assert.

File: lib/inspector/inspector_controller.dart [hunk 5]
```diff
@@ -322,8 +321,12 @@ class InspectorController implements InspectorServiceClient {
       _treeGroups.promoteNext();
       clearValueToInspectorTreeNodeMapping();
       if (node != null) {
-        final InspectorTreeNode rootNode =
-            setupInspectorTreeNode(inspectorTree.createNode(), node, true);
+        final InspectorTreeNode rootNode = inspectorTree.setupInspectorTreeNode(
+          inspectorTree.createNode(),
+          node,
+          expandChildren: true,
+          expandProperties: false
```
Comment (jacob314) [inline lib/inspector/inspector_controller.dart]: add trailing comma.

File: lib/logging/logging.css [hunk 6]
```diff
@@ -13,3 +13,24 @@ td.log-label-column {
     padding-left: 5px;
     padding-right: 5px;
 }
+
+.logging-container {
```
Comment (devoncarew) [inline lib/logging/logging.css]: We may be able to get by with slightly fewer properties here and for `.logging-panel` (I think we can rely on some of the existing flexbox attributes we've defined). Happy to patch this in locally and try and minimize the explicit width / height usages here.

File: lib/logging/logging.dart [hunk 7]
```diff
@@ -168,6 +180,20 @@ class LoggingScreen extends Screen {
         e.timestamp,
         summaryHtml: '$frameInfo$div',
       ));
+    } else if (e.extensionKind == 'Flutter.Error') {
+      final RemoteDiagnosticsNode node =
+          RemoteDiagnosticsNode(e.extensionData.data, objectGroup, false, null);
+      if (_verboseDebugging) {
```
Comment (jacob314) [inline lib/logging/logging.dart]: remove XXX from this code.
I use XXX to indicate code that should be reviewed before checking in.

File: lib/logging/logging.dart [hunk 8]
```diff
@@ -550,6 +578,8 @@ String getCssClassForEventKind(LogData item) {
     cssClass = 'stderr';
   } else if (item.kind == 'stdout') {
     cssClass = 'stdout';
+  } else if (item.kind == 'flutter.error') {
+    cssClass = 'stderr';
```
Comment (jacob314) [inline lib/logging/logging.dart]: nice!

File: lib/logging/logging.dart [hunk 9]
```diff
@@ -578,12 +608,57 @@ class LogDetailsUI extends CoreElement {
...
+      message.clear();
+      tree = InspectorTreeHtml(
+        summaryTree: false,
+        treeType: FlutterTreeType.widget,
+        onSelectionChange: () {
+          final InspectorTreeNode node = tree.selection;
+          if (node != null) {
+            tree.maybePopulateChildren(node);
+          }
+          // The ui wasn't updated because this selection was triggered by the logging
```
Comment (jacob314) [inline lib/logging/logging.dart]: Remove this comment. it is somewhat wrong.

File: lib/inspector/inspector_controller.dart (hunk 5 above)
Comment (pq) [reply lib/inspector/inspector_controller.dart]: Done.

File: lib/logging/logging.dart (hunk 3 above)
Comment (jacob314) [inline lib/logging/logging.dart]: remove this print statement. Consider removing both these methods as they aren't really useful for this view.

File: lib/inspector/diagnostics_node.dart (hunk 1 above)
Comment (pq) [reply lib/inspector/diagnostics_node.dart]: 👍  Done.

File: lib/logging/logging.dart [hunk 10]
```diff
@@ -578,12 +608,57 @@ class LogDetailsUI extends CoreElement {
...
+        data.node,
+        expandChildren: true,
+        expandProperties: true,
+      );
+      // No sense in collapsing the root node.
+      root.allowExpandCollapse = false;
+      tree.root = root;
+      message.add(tree.element);
+
+      if (_verboseDebugging) {
```
Comment (jacob314) [inline lib/logging/logging.dart]: Delete this block. Adding this div slowed down things too much.

File: lib/timeline/timeline.dart [hunk 11]
```diff
@@ -62,11 +61,7 @@ class TimelineScreen extends Screen {
     final CoreElement upperButtonSection = div(c: 'section')
       ..layoutHorizontal()
       ..add(<CoreElement>[
-        div(c: 'btn-group')
-          ..add([
-            createHotReloadButton(),
-            createHotRestartButton(),
-          ]),
+        createHotReloadRestartGroup(framework),
```
Comment (jacob314) [inline lib/timeline/timeline.dart]: @devoncarew 
here's the hot reload logic refactored out slightly.

File: lib/inspector/diagnostics_node.dart (hunk 2 above)
Comment (pq) [reply lib/inspector/diagnostics_node.dart]: 👍

Comment (pq) [reply lib/inspector/diagnostics_node.dart]: Updated assert as per @jacob314's suggestion.

**Comment by jacob314** [comment]
> Is your screenshot up to date? I don't see the code to create the reload and restart button in logging.dart (and generally, think that we shouldn't put them in every page, but in some future PR create a area in the UI that's global to the app, which can hold those buttons).
Screenshot assumes a refactor on the package:flutter side which hasn't landed yet.

File: lib/logging/logging.dart (hunk 3 above)
Comment (pq) [reply lib/logging/logging.dart]: Agreed.  Fixed.

File: lib/ui/fake_flutter/diagnosticable.dart (hunk 4 above)
Comment (pq) [reply lib/ui/fake_flutter/diagnosticable.dart]: I think not useful.  Removing.

File: lib/logging/logging.dart (hunk 7 above)
Comment (pq) [reply lib/logging/logging.dart]: 👍

File: lib/logging/logging.dart (hunk 9 above)
Comment (pq) [reply lib/logging/logging.dart]: 👍

File: lib/logging/logging.dart (hunk 3 above)
Comment (pq) [reply lib/logging/logging.dart]: Removed methods.

File: lib/logging/logging.dart (hunk 10 above)
Comment (pq) [reply lib/logging/logging.dart]: 👍

File: lib/logging/logging.dart (hunk 7 above)
Comment (pq) [reply lib/logging/logging.dart]: Done.

**Comment by pq** [comment]
As per discussion w/ @jacob314, landing and he'll follow-up on the inspector test timeouts.

**Pull request merged**